### DISCUSSION TITLE
Required vs. optional dependencies

### DISCUSSION THREAD

**Discussion by adamjstewart** [torchgeo/torchgeo#662] [General]
My goal is to make `pip install torchgeo` somewhat lightweight while still enabling most features by default. We do this by maintaining a set of "required" dependencies and a separate set of "optional" dependencies. `pip install torchgeo` only installs the former, while `pip install torchgeo[datasets]` installs everything.

I wanted to go through our current dependencies and see if any should be moved around. Here is how many features require each dependency:

### Required (currently)

* einops (2 datamodules, 1 model)
* fiona (6 datasets)
* kornia (4 datamodules, 1 trainer, 1 transform)
* matplotlib (all datasets, all datamodules)
* numpy (40 datasets)
* omegaconf (`train.py`, trainer tests)
* packaging (1 dataset, 1 model, 2 trainers)
* pillow (20 datasets)
* pyproj (2 datasets)
* pytorch-lightning (all datamodules, all trainers)
* rasterio (30 datasets)
* rtree (all GeoDatasets, GeoSamplers)
* scikit-learn (2 datamodules)
* segmentation-models-pytorch (1 model, 2 trainers)
* shapely (3 datasets)
* timm (1 trainer)
* torch (the whole library)
* torchmetrics (all trainers)
* torchvision (the whole library)

### Optional (currently)

* h5py (2 datasets)
* laspy (1 dataset)
* open3d (1 dataset)
* opencv (1 dataset)
* pandas (7 datasets)
* pycocotools (1 dataset)
* radiant-mlhub (5 datasets)
* rarfile (2 datasets)
* scipy (1 dataset)
* zipfile-deflate64 (1 dataset)

### Considerations

1. If a dependency is already required by a required dependency, we might as well make it required
2. If a dependency is a very popular ML/geospatial library, we might as well make it required
3. If a dependency is only needed for limited functionality, we should make it optional

### Results

I think einops and omegaconf should definitely be optional, they are barely used in TorchGeo. Omegaconf can be moved under the `tests` group of dependencies. If we ever add entrypoints for our top-level scripts, we may want to reconsider.

I think pandas should definitely be required, it's used by a lot of datasets and is a core ML library. There are two more datasets that use `csv` that could be ported to pandas.

I'm on the fence about pyproj, scikit-learn, scipy, and shapely. All are barely used in TorchGeo, but they're also core ML/geospatial libraries that most users will already need anyway.

I'm also on the fence about segmentation-models-pytorch and timm. I think we'll likely use those in more places and they'll become core to our trainer infrastructure, so we may want to keep them as required.

We should consider replacing open3d with pyvista. Open3d is a nightmare to install and lacks Python 3.10 wheels. The tests fail on macOS with mysterious seg faults. Pyvista will also be a nightmare to install, but seems better supported. It's a pure Python library, but requires VTK. VTK also lacks Python 3.10 wheels though. @isaaccorley will investigate this.

Should we rename the "datasets" extras group to "optional"? I guess it depends on how many of our optional dependencies are used outside of datasets/datamodules.

Continuation of #72

**Comment by isaaccorley** [comment]
My only suggestion is that we should be using einops more since it's more readable and provides names to dimensions, e.g. BCHW. It also has no dependencies (only depends on torch during runtime).
